assets/src/components/v2/persistent_carousel.tsx

```diff
@@ -0,0 +1,99 @@
+import React, { ComponentType, ReactNode, useEffect, useState } from "react";
+import makePersistent, { WrappedComponentProps } from "./persistent_wrapper";
+
+interface PageRendererProps<T> {
+  page: T;
+  pageIndex: number;
+  numPages: number;
+}
+
+interface Props<T> extends WrappedComponentProps {
+  PageRenderer: ComponentType<PageRendererProps<T>>;
+  pages: T[];
+}
+
+const Carousel = <T,>({ PageRenderer, pages, onFinish, lastUpdate }: Props<T>): ReactNode => {
+  const [isFirstRender, setIsFirstRender] = useState(true);
+  const [pageIndex, setPageIndex] = useState(0);
+
+  useEffect(() => {
+    if (lastUpdate != null) {
+      if (isFirstRender) {
+        setIsFirstRender(false);
+      } else {
+        setPageIndex((i) => i + 1);
+      }
+    }
+  }, [lastUpdate]);
+
+  useEffect(() => {
+    if (pageIndex === pages.length - 1) {
+      onFinish();
+    }
+  }, [pageIndex]);
+
+  return (
+    <PageRenderer page={pages[pageIndex]} pageIndex={pageIndex} numPages={pages.length} />
+  );
+}
+
+const PersistentCarousel = makePersistent(Carousel as ComponentType<WrappedComponentProps>);
+
+/**
+ * Call this function on some `PageRenderer` component to wrap it in a persistent component that
+ * expects a `pages` prop containing an array of data, and will pass one element of the array at a time to
+ * `PageRenderer`.
```

Comment:
Is there a reason this is formatted like this instead of as a @moduledoc?

Reply:
Yup, this is just standard JSdoc which I think is borrowed from Javadoc? `@moduledoc` is Elixir's syntax for essentially the same thing, so we use that to attach documentation to modules in the backend code.